pkg/controller/license/trial/trial_controller.go

```diff
@@ -80,85 +78,95 @@ func (r *ReconcileTrials) Reconcile(request reconcile.Request) (reconcile.Result
+	var trialStatus corev1.Secret
+	err := r.Get(types.NamespacedName{Namespace: r.operatorNamespace, Name: licensing.TrialStatusSecretKey}, &trialStatus)
+	if errors.IsNotFound(err) {
+		if !r.trialState.IsTrialRunning() {
+			// we have no state in memory nor in the status secret: start the activation process
+			if err := r.startTrialActivation(); err != nil {
+				return err
+			}
+		}
 
-func (r *ReconcileTrials) initTrial(secret corev1.Secret, l licensing.EnterpriseLicense) error {
-	if r.isTrialRunning() {
-		setValidationMsg(&secret, trialOnlyOnceMsg)
-		return licensing.UpdateEnterpriseLicense(r, secret, l)
+		// we have state in memory but the status secret is missing: recreate it
+		trialStatus, err = licensing.ExpectedTrialStatus(r.operatorNamespace, license, r.trialState)
+		if err != nil {
+			return fmt.Errorf("while creating expected trial status %w", err)
+		}
+		return r.Create(&trialStatus)
 	}
-
-	trialPubKey, err := licensing.InitTrial(r, r.operatorNamespace, secret, &l)
 	if err != nil {
-		return err
+		return fmt.Errorf("while fetching trial status %w", err)
 	}
-	// retain pub key in memory for later iterations
-	r.trialPubKey = trialPubKey
-	return nil
-}
 
-func (r *ReconcileTrials) reconcileTrialStatus(trialStatus corev1.Secret) error {
-	if !r.isTrialRunning() {
-		// reinstate pubkey from status secret e.g. after operator restart
-		pubKeyBytes := trialStatus.Data[licensing.TrialPubkeyKey]
-		key, err := licensing.ParsePubKey(pubKeyBytes)
+	// the status secret is there but we don't have anything in memory: recover the state
+	if r.trialState.IsEmpty() {
+		recoveredKeys, err := licensing.NewTrialStateFromStatus(trialStatus)
 		if err != nil {
 			return err
 		}
-		r.trialPubKey = key
-		return nil
+		r.trialState = recoveredKeys
 	}
-	pubkeyBytes, err := x509.MarshalPKIXPublicKey(r.trialPubKey)
+	// if trial status exists, but:
+	// - has been tampered with: reconstruct it
+	// - we need to update it to complete the trial activation
+	// - we need to update it because we just regenerated the state after a crash
+	expected, err := licensing.ExpectedTrialStatus(r.operatorNamespace, license, r.trialState)
 	if err != nil {
 		return err
 	}
-	if bytes.Equal(trialStatus.Data[licensing.TrialPubkeyKey], pubkeyBytes) {
+	if reflect.DeepEqual(expected.Data, trialStatus.Data) {
 		return nil
 	}
-	if trialStatus.Data == nil {
-		trialStatus.Data = map[string][]byte{} // if trial status has been tampered with
-	}
-	trialStatus.Data[licensing.TrialPubkeyKey] = pubkeyBytes
+	trialStatus.Data = expected.Data
 	return r.Update(&trialStatus)
+}
+
+func (r *ReconcileTrials) startTrialActivation() error {
+	state, err := licensing.NewTrialState()
+	if err != nil {
+		return err
+	}
+	r.trialState = state
+	return nil
+}
 
+func (r *ReconcileTrials) completeTrialActivation() (reconcile.Result, error) {
+	if r.trialState.CompleteTrialActivation() {
+		// requeue to update trial status
+		return reconcile.Result{Requeue: true}, nil
```

Comment:
why don't we update the status at this reconciliation instead of requeuing?

Reply:
We can. The reason I did not do it initially was that I thought we might run into conflicts when we update twice in the same reconciliation. Also it felt easier to reason about the code when all status updates happen in `reconcileStatus`